pal/AzureIoTSocket_Ethernet/socketio_esp32ethernet2.cpp

```diff
@@ -0,0 +1,577 @@
+        free(socket_io);
+    }
+}
+
+int socketio_open(CONCRETE_IO_HANDLE socket_io, ON_IO_OPEN_COMPLETE on_io_open_complete, void* on_io_open_complete_context, ON_BYTES_RECEIVED on_bytes_received, void* on_bytes_received_context, ON_IO_ERROR on_io_error, void* on_io_error_context)
+{
+    int result = 0;
+
+    SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+    if (socket_io == NULL)
+    {
+        LogError("Invalid argument: SOCKET_IO_INSTANCE is NULL");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        if (socket_io_instance->io_state != IO_STATE_CLOSED)
+        {
+            LogError("Failure: socket state is not closed.");
+            result = MU_FAILURE;
+        }
+        else if (socket_io_instance->socket != NULL)
+        {
+            // Opening an accepted socket
+            socket_io_instance->on_bytes_received_context = on_bytes_received_context;
+            socket_io_instance->on_bytes_received = on_bytes_received;
+            socket_io_instance->on_io_error = on_io_error;
+            socket_io_instance->on_io_error_context = on_io_error_context;
+            socket_io_instance->io_state = IO_STATE_OPEN;
+
+            result = 0;
+        }
+        else
+        {
+            socket_io_instance->socket = new EthernetClient();
+
+            if (socket_io_instance->socket == NULL)
+            {
+                LogError("Failure: socket create failure %d.", socket_io_instance->socket);
+                result = MU_FAILURE;
+            }
+            else if (!socket_io_instance->socket->connect(socket_io_instance->hostname, socket_io_instance->port) != 0)
+            {
+                LogError("Socket connect failed");
+                result = MU_FAILURE;
+            }
+
+            if (result == 0)
+            {
+                socket_io_instance->on_bytes_received = on_bytes_received;
+                socket_io_instance->on_bytes_received_context = on_bytes_received_context;
+
+                socket_io_instance->on_io_error = on_io_error;
+                socket_io_instance->on_io_error_context = on_io_error_context;
+
+                socket_io_instance->io_state = IO_STATE_OPEN;
+            }
+            else
+            {
+                if (socket_io_instance->socket != NULL)
+                {
+                    delete socket_io_instance->socket;
+                    socket_io_instance->socket == NULL;
+                }
+            }
+        }
+    }
+
+    if (on_io_open_complete != NULL)
+    {
+        on_io_open_complete(on_io_open_complete_context, result == 0 ? IO_OPEN_OK : IO_OPEN_ERROR);
+    }
+
+    return result;
+}
+
+int socketio_close(CONCRETE_IO_HANDLE socket_io, ON_IO_CLOSE_COMPLETE on_io_close_complete, void* callback_context)
+{
+    int result = 0;
+
+    if (socket_io == NULL)
+    {
+        result = MU_FAILURE;
+    }
+    else
+    {
+        SOCKET_IO_INSTANCE* socket_io_instance = (SOCKET_IO_INSTANCE*)socket_io;
+		
```

Comment:
still have some tabs here

Reply:
just as a helpful tip you can use the VSCode regex search option (far right .* button on the find) and search for `\t` for all tab instances.

Reply:
Done.